### DISCUSSION TITLE
Some ideas relating to the procedural generation of visual elements associated with notes/concepts

### DISCUSSION THREAD

**Discussion by khughitt** [flowershow/flowershow#70] [General and Help]
Greetings!

This looks like a really cool project! Thanks for making it open-source from the get-go!

I just wanted to share a few ideas relating to the possible application of viz/proc-gen, discussed briefly with @rufuspollock.

_Basic thoughts_:

1. It could be useful to associate each note (or topic/entity in an eventually knowledgebase implementation) with some visual element.
2. In addition to just making for a (potentially) more fun and playful experience, persistent visualizations associated with commonly-considered ideas can be a helpful way to reinforce the memory.
3. 2d/3d procedural generation could be used to generate such visual entities. [three.js](https://threejs.org/) is great for both of these.
4. If one can divide notes up into meaningful semantic groups (e.g. via clustering of document embeddings), then a color could be assigned to each semantic group and used for the procedurally-generated elements.
5. There are different types of visual elements one could imagine generating (e.g. [trees](https://mdh.diva-portal.org/smash/get/diva2:1105996/FULLTEXT01.pdf), [planets](https://stroemer.cc/procedural-generation-gas-giants/), [etc](https://github.com/LingDong-/fishdraw).), but, given the theme of this tool, a good candidate might be to procedurally generate flowers and other plants as part of a garden. 
  - ex. [here](([this](https://www.reddit.com/r/Art/comments/ukxsse/spring_me_digital_2022/) ) (not procedurally-generated afaik, but a pretty reference example)
6. One could further imagine "growing" plants/flowers in the garden in proportion to the amount of activity the note/topic gets.

Some of these ideas might not be relevant at all to your goals, but it's something to consider, perhaps. :tulip: 

Keep up the great work!
Keith

**Comment by rufuspollock** [comment]
@khughitt welcome, thanks for posting!

Really like this idea.

> It could be useful to associate each note (or topic/entity in an eventually knowledgebase implementation) with some visual element.

Great idea. Would this be an emoji or image or something else. Would it be auto-generated? (You mention that further down).

I really like the idea of autogenerating plants in some way 😄

🌹 🌷 🥀 💐

**Comment by khughitt** [comment]
You could give the user some control over the image/3d model generation process, but I think it's interesting to think about creating something that is automatic by default.

Each note is associated with a random seed that determines some unique set of parameters for the image/model construction algorithm.

If you want to start simple, you could consider starting with something abstract but (possibly?) memorable, e.g. a circle parameterized by:

- color1
- color2
- radius
- noise seed
- opacity1
- opacity2
- etc.

By itself, this is perhaps not so interesting or useful, but if you were to cluster your notes into a few different topics, then you could start to fix "color1" ~ cluster, and perhaps use color2 to indicate secondary cluster membership or some other contextual information, if there are some categories of notes that frequently reappear and differ by context instead of context (e.g. "todo", "reminders", "ideas on xx", etc.).

For realistic generation of plants, check out [Algorithmic Beauty of Plants](http://algorithmicbotany.org/papers/abop/abop.pdf) by Prusinkiewicz and Lindenmeyer. It describes a recursive string-based approach to simulated plants and plant growth called 
 [L-systems](https://en.wikipedia.org/wiki/L-systems) (named after one of the authors, "Lindenmeyer").

**Reply by rufuspollock** [reply]
Great suggestions @khughitt and love the link to algorithmic beauty stuff.
